app/screens/image_preview/downloader.ios.js

```diff
@@ -76,82 +97,149 @@ export default class Downloader extends PureComponent {
+
+                options.path = `${downloadPath}/${file.id}.${file.extension}`;
+            } else {
+                options.fileCache = true;
+                options.appendExt = file.extension;
+            }
+
+            this.downloadTask = RNFetchBlob.config(options).fetch('GET', imageUrl).
+                progress((received, total) => {
+                    const progress = (received / total) * 100;
+                    if (this.mounted) {
+                        this.setState({
+                            progress, started: true
+                        });
+                    }
                 });
-            });
 
             const res = await this.downloadTask;
-            const path = res.path();
-            const newPath = await CameraRoll.saveToCameraRoll(path, 'photo');
+            let path = res.path();
 
-            this.setState({
-                progress: 100
-            }, () => {
-                // need to wait a bit for the progress circle UI to update to the give progress
-                setTimeout(async () => {
-                    try {
-                        // handles the case of a late cancellation by the user
-                        // and ensures that we remove the file if they did cancel late.
-                        if (this.state.didCancel) {
-                            // TODO: There's issue with deleting files from the cameraRoll on iOS here https://github.com/wkh237/react-native-fetch-blob/issues/479
-                            // This only occurs if the user cancels when the download is around 80% or more
-                            await RNFetchBlob.fs.unlink(newPath);
+            if (saveToCameraRoll) {
+                path = await CameraRoll.saveToCameraRoll(path, 'photo');
+            }
+
+            if (this.mounted) {
+                this.setState({
+                    progress: 100
+                }, () => {
+                    // need to wait a bit for the progress circle UI to update to the give progress
+                    setTimeout(async () => {
+                        try {
+                            // handles the case of a late cancellation by the user
+                            // and ensures that we remove the file if they did cancel late.
+                            if (this.state.didCancel) {
+                                // TODO: There's issue with deleting files from the cameraRoll on iOS here https://github.com/wkh237/react-native-fetch-blob/issues/479
+                                // This only occurs if the user cancels when the download is around 80% or more
+                                await RNFetchBlob.fs.unlink(path);
+                                this.props.onDownloadCancel();
+                            } else {
+                                this.props.onDownloadSuccess();
+                            }
+                        } catch (error) {
+                            // ensure the downloader at least closes if a error
```

Comment:
This could be done in a finally like
```
if (this.state.didCancel) {
    try {
        await RNFetchBlob.fs.unlink(path);
    } finally {
        this.props.onDownloadCancel();
    }
} else {
    this.props.onDownloadSuccess();
}
```
to get rid of one of the onDownloadCancel calls